### ISSUE TITLE
Inquiry Regarding Model Training and Generation Correlation

### ISSUE THREAD

**Issue by ldhkc** [atomistic-machine-learning/schnetpack-gschnet#17]
Dear Dr. Niklas Gebauer,

I hope this message finds you well.

I have encountered a few points of confusion regarding your model and would greatly appreciate your insights.

Specifically, I am wondering whether there is any correlation between the training phase and the generation phase of the model. More precisely, do the various loss metrics during training (such as test_loss, test_loss_dist, test_loss_type, train_loss, val_loss, val_loss_dist, val_loss_type, etc.) need to be minimized for the model's generation performance to improve? By "generation performance," I mean the metrics obtained after running the check_validity.py script, such as the number of valid molecules, the number of unique molecules, etc., which correspond to the model generating a specified number of molecules.

During my experiments, I encountered an interesting phenomenon that I find puzzling: in one experiment, the training loss metrics (including test and validation losses) were lower compared to the original experiment, yet the generation results were not as good as those from the original experiment. Could you provide any possible explanations or reasons for this discrepancy?

I look forward to your guidance, and I greatly appreciate your time and help.

Best regards,
ldhkc

**Comment by NiklasGebauer** [comment]
Dear @ldhkc ,

I'm sorry for the very delayed reply. I hope my answers might still be helpful for you:

In my experience, there is a clear correlation in the beginning of the training. As the losses go down, the model will start to be able to sample valid molecules. However, as you have observed, it can happen that at some point further lowering the loss will lead to worse samples with respect to specific metrics:
 1. The most intuitive example is if your metric is the number of novel molecules - we observed that sometimes the number of novel molecules decreases as the loss decreases, i.e. the model starts to sample more molecules from the training set. This could be interpreted as some kind of overfitting. However, at the same time, the number of valid molecules might still be increasing. Therefore, we might have a trade-off between the validity (sampling valid structures) and the diversity (sampling novel/diverse structures). I think this trade-off is somewhat inherent to generative models and finding the most suitable hyperparameters for your application depends on many things: The dataset (e.g. size of molecules, number of samples), the model (number of parameters), the loss scheduling, and of course the metric you are interested in.
 2. There is also a factor that is specific to our architecture: During training, we only optimize to predict the correct distance to the new atom. However, during sampling, we construct a 3d distribution which we compute from the predicted distances. This step is not part of the training. The positive side of this is that training is comparatively fast and straight forward. The negative side is that the loss (on distances) does not directly represent accuracy of the 3d distribution. While I would argue that getting the distances correct usually leads to better 3d distributions as well, we know that there can exist artifacts (i.e. probability mass at unreasonable positions) when constructing the distribution only from pairwise distances. Maybe you ended up with a model that for some reason (which is not intuitive to me) suffers from more artifacts in the construction of the 3d distribution. (If you are interested in more details on the reconstruction artifacts, you could read section 3.1.1 of my [thesis](https://depositonce.tu-berlin.de/items/5e3896d7-334b-47a6-9b0d-e8912abeca8f).)
 3. Another possible explanation is that the model gets more susceptible to misplaced atoms as it more properly captures the training data (or starts to overfit the training set). Imagine that during the generation, the fifth atom is sampled at a position which is in some sense unrealistic. I suspect the model with lower loss might tend to a more extreme change in the prediction of the next atom, leading to broken structures (i.e. it might not tolerate deviations in distances in the input as well as the less accurate model with higher loss).
 4. There is also another trade-off involved. Our loss is computed as loss=loss_dist+loss_type. I observed that sometimes at late stages of the training, the prediction of one of these two gets better while the other one gets worse. It could be that one of them is more important to optimize the metric you are looking at and therefore the model with lower loss actually gets worse at that metric.

These are just some ideas of why such observations of the model getting "worse" at sampling could occur. Of course we would need to thoroughly test different models and look at the sampling procedure, predicted distances, constructed 3d distributions etc. in order to assess what is happening.

Please let me know in case you have further questions or need clarifications on some of these points.

Kind regards,
Niklas

**Comment by ldhkc** [comment]
@NiklasGebauer Thank you very much for your detailed reply. After reading it, I now have a general understanding of the relevant content. If I have any other questions later on, I will consult you again.
